nexusadspy/segment.py

```diff
@@ -0,0 +1,100 @@
+
+from __future__ import (
+    print_function, division, generators,
+    absolute_import, unicode_literals
+)
+
+from io import BytesIO
+from gzip import GzipFile
+import time
+import logging
+
+from nexusadspy.client import AppnexusClient
+
+
+class SegmentsUploader:
+
+    BATCH_UPLOAD_ANDROID_SPECIFIER = "8"
+    BATCH_UPLOAD_IOS_SPECIFIER = "3"
+
+    def __init__(self, users_list, segment_code, separators, member_id,
+                 credentials_path='.appnexus_auth.json'):
+        """
+        Initialize batch-upload API wrapper for appnexus.
+        :param users_list: list, List of dictionaries representing appnexus users. Every member should have fields
+            -uid: Appnexus user ID. AAID/IDFS in case of mobile
+            -timestamp: Timestamp when user entered the segment
+            -expiration (optional): Expiration timestamp for the user. Default 0.
+            -value (optional): Numerical value for the segment. Default 0.
+            -mobile_os (optional): OS used by the user. Considered to be desktop if absent.
+        :param segment_code: str, Segment code to add users in.
+        :param separators: array, Array of five field separators. As documented in 
+        https://wiki.appnexus.com/display/api/Batch+Segment+Service+-+File+Format#BatchSegmentService-FileFormat-Separators
+        :param member_id: str, Member ID for appnexus account.
+        :param credentials_path: str, Credentials path for AppnexusClient.
+        :return:
+        """
+        self._credentials_path = credentials_path
+        self._users_list = users_list
+        self._segment_code = segment_code  # Appnexus bug: Segment upload batch API does not work with segment IDs
+        self._separators = separators
+        self._member_id = member_id
+        self._job_id = None
+        self.logger = logging.getLogger('AppnexusClient.segment')
+
+    def upload(self, polling_duration_sec=2):
+        """
+        Initiate segment upload task
+        :param polling_duration_sec: int, Time to sleep while polling for status
+        :return: tuple, with two values, number of valid users and invalid users
+        """
+        valid_user_count = invalid_user_count = 0
+        api_client = AppnexusClient(self._credentials_path)
+        self._initialize_upload(api_client)
+        while True:
+            time.sleep(polling_duration_sec)
+            job_status = self._get_job_status_response(api_client)
+            if job_status[0].get('phase') == 'completed':
+                valid_user_count = job_status[0]['batch_segment_upload_job']['num_valid_user']
+                invalid_user_count = job_status[0]['batch_segment_upload_job']['num_invalid_user']
+                break
+        return valid_user_count, invalid_user_count
+
+    def _initialize_upload(self, api_client):
```

Comment:
this method does two things: request a job ID and post the upload data. split this method into two methods.